### ISSUE TITLE
Possibility to modify JenaModel before response is returned by JMH/JaxRsProviders

### ISSUE THREAD

**Issue by jadelkhoury** [eclipse/lyo#76]
Can we introduce a mechanism that allows the devleoper to manipulate the response returned by LyoCore?

This would be done by either allowing custom functionality in JenaHelper, or AbstractOslcRdfXmlProvider?

https://github.com/eclipse/lyo/blob/b5f7207f969b4d31f0dc52756bf61229110e75b3/core/oslc4j-jena-provider/src/main/java/org/eclipse/lyo/oslc4j/provider/jena/AbstractOslcRdfXmlProvider.java#L146

@berezovskyi @jamsden! Any tips on how we can do this elegantly?

**Comment by jadelkhoury** [comment]
How about the following proposal?

1. We introduce a method "visit(model, ....)" under AbstractResource (or IExtendedResource?).
2. At the end of the createJenaModel method (https://github.com/eclipse/lyo/blob/b5f7207f969b4d31f0dc52756bf61229110e75b3/core/oslc4j-jena-provider/src/main/java/org/eclipse/lyo/oslc4j/provider/jena/JenaModelHelper.java#L212), JMH loops through its "objects" and calls visit(model).
3. Resource Classes can then override the visit method to modify the final model as desired. 
4. We will need a similar method at end of fromJenaResource() as well.

**Comment by berezovskyi** [comment]
(2) is very much not thread safe. please do not any more static stuff to that class.

(1) Visitor pattern is in my opinion not needed here and JAX-RS interceptors are a much better fit https://docs.jboss.org/resteasy/docs/4.0.0.Final/userguide/html/Interceptors.html#d4e2053

**Comment by jamsden** [comment]
I'm left with a nagging feeling about data integrity and what actually constitutes a resource representation. Are we crossing into the ETL domain? Can clients have a reasonable expectation about the provenance of the resources? Maybe I'm overthinking this.

**Comment by berezovskyi** [comment]
Right, JAX-RS interceptors would also allow Lyo not to step into that area either. But I agree with Jad that such hacks are very much needed in the OSLC space, sadly, and Lyo should at least provide for a way to get the job done.

**Comment by berezovskyi** [comment]
Practical advice (did not try, but that's how I would do it):

1. Implement https://docs.oracle.com/javaee/7/api/javax/ws/rs/ext/WriterInterceptor.html
2. Call https://docs.oracle.com/javaee/7/api/javax/ws/rs/ext/WriterInterceptorContext.html#getEntity--, cast it to an AbstractResource or its array.
3. Do your thing, potentially calling JMH and converting objects to a Jena Model and editing it on a triple level.
4. Call https://docs.oracle.com/javaee/7/api/javax/ws/rs/ext/WriterInterceptorContext.html#getHeaders-- and set any extra headers as needed.
5. Call https://docs.oracle.com/javaee/7/api/javax/ws/rs/ext/WriterInterceptorContext.html#setEntity-java.lang.Object-. If you received an array of JMH POJOs and here you set a Jena Model, this should even change a MessageBodyWriter instance used.
6. Call https://docs.oracle.com/javaee/7/api/javax/ws/rs/ext/WriterInterceptorContext.html#proceed--

**Comment by berezovskyi** [comment]
We may want to revive https://github.com/eclipse/lyo.core/pull/76/files#diff-0f7af1eb04acfdfbffe4b21db24c1c2c0ba064f2a1d537d6a763bda00afe6166 and also add a standard writer for a Jena Model.

**Comment by jadelkhoury** [comment]
I looked into JAX-RS interceptors. The problem there is that one will receive a stream, and then one will have to reconstruct the Jena model before it can be modified. I wanted to catch the return results as a Jena model.

@Jim! Not sure I understood your reservation, because this is not related to client/server. Sometimes it is just not enough to rely on the annotations to tell JMH what to do. A good example is the SysmlServer, where we wanted to perform some changes to the Jena model. For example, adding a new property, or changing a value of a property.
It would be nice if these can be done more systematically. 

In my new case, I want to serialize/deserialize an object in a custom way, and I can't do that using annotations.

**Comment by jadelkhoury** [comment]
but point taken about concurrency. I will look into your many pointers.

**Comment by berezovskyi** [comment]
Jad, I assume there should be a way to intercept a JAX-RS response before the streams are involved. Let's try to have a Zoom session quickly to check this.

**Comment by jadelkhoury** [comment]
@berezovskyi 

The Jax-RS Provider writes to the stream. so unless we modify the JenaProviders we have built, I can't see how the Jax-RS Interceptors can work to embed code in the middle of the Provider.

**Comment by berezovskyi** [comment]
https://github.com/oslc-op/refimpl/pull/27
